File: Algorithm_538_Convert BST to Greater Tree.cpp
#include <iostream>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
public:
TreeNode *sumRoot;
    TreeNode* convertBST(TreeNode* root) {
        sumRoot=getSumTree(root);
        TreeNode *cur=root, *s=sumRoot, *pre=NULL;
        add(root,sumRoot,pre,0);
        return root;
    }
    TreeNode* getSumTree(TreeNode* root)
    {
        if (!root) return NULL;
        TreeNode *newRoot=new TreeNode(0);
        int leftSum=0,rightSum=0;
        if (root->left)
        {
            newRoot->left=getSumTree(root->left);
            leftSum=newRoot->left->val;
        }
        if (root->right)
        {
            newRoot->right=getSumTree(root->right);
            rightSum=newRoot->right->val;
        }
        newRoot->val=root->val+leftSum+rightSum;
        return newRoot;
    }
    void add(TreeNode *cur, TreeNode *s, TreeNode *pre, int sum)
    {
        if (!cur) return;
        if (pre && pre->left==s) sum+=pre->val-s->val;
        cur->val+=sum+(s->right?s->right->val:0);
        add(cur->left,s->left,s,sum);
        add(cur->right,s->right,s,sum);
    }
};
/*class Solution {
public:
    int sum=0;
    TreeNode* convertBST(TreeNode* root)
    {
        if (!root) return NULL;
        if (root->right) convertBST(root->right);
        sum+=root->val;
        root->val=sum;
        if (root->left) convertBST(root->left);
        return root;
    }
};*/

File: Algorithm_538_Convert BST to Greater Tree.java
// 思路：变量sum存储所有比root大的节点值之和
// 由于BST的右子树所有节点值全部大于root，因此先对右子树递归
// 对右子树的递归完成后，把sum加到当前root.val上，并把新的root.val作为新的sum
// 由于左子树所有节点值全部小于root，因此以上操作完成后再对左子树递归

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class Solution {
    
    private int sum = 0;
    
    public TreeNode convertBST(TreeNode root) {
        helper(root);
        return root;
    }

    public void helper(TreeNode root) {
        if (root == null) {
            return;
        }
        helper(root.right);
        root.val += sum;
        sum = root.val;
        helper(root.left);
    }

}